### ISSUE TITLE
Parenthesis matching in markdown cells

### ISSUE THREAD

**Issue by mocquin** [jupyterlab/jupyterlab#5892]
When editing a markdown cell in a notebook, it seems that there is no parenthesis matching, while this feature is enabled when editing a markdown file.

## Editing a markdown file
![latex_1](https://user-images.githubusercontent.com/36188368/51459902-4a296600-1d5a-11e9-9fe3-f0f0c21bfd3b.jpg)
(parenthesis matching highlighting)

## Editing a markdown cell
![latex_2](https://user-images.githubusercontent.com/36188368/51459938-67f6cb00-1d5a-11e9-9c67-5c205040b55d.jpg)
(no parenthesis matching highlighting)

I think it'd be great to have this feature in markdown cells as well ;)

**Comment by telamonian** [comment]
This option already exists, but it's a little buried. Here's how to turn it on: 

- start up Jupyterlab
- go the `Settings` menu
- select `Advanced Settings Editor`
- on the left hand side of the settings editor, select `Notebook`
- this will open up two columns that will look something like this:

![image](https://user-images.githubusercontent.com/2263641/51476406-74c3f080-1d53-11e9-972a-6d0b332b8423.png)

- in the left column are the default settings, in a JSON format. Copy and paste these over the right column (which stores the user settings) and modify them how you like
- there are three groups of options, one for each of the notebook cell types (code, markdown, and raw). You're looking for the `matchBrackets` option in the `markdownCellConfig` group. In the right hand column change the value of this option from `false` to `true`.

We should really make that easier to do. Honestly, I had a bit of trouble with formatting the JSON correctly the first time I used the settings editor.

**Comment by mocquin** [comment]
Thanks a lot works as expected !

Concerning this feature ("matchBrackets"), I'm wondering if there is a reason it defaults to false ?
Concerning the JSON settings, should we "highlight" the diff between the system defaults and the user overrides ? It'd be convenient to check what are the settings the user modified.

**Comment by telamonian** [comment]
I think the simplest thing would be to make the "Table View" in the current settings editor actually, you know, do something. That would probably be the easiest way to implement something significantly more useable than what we currently have.

It would be straightforward to design for the case when a category's options are a flat list. For the case of a nested hierarchy of options (like with Notebook), maybe there could be something like a filebrowser widget in the left hand column. Do you know of any extensions with a version of the filebrowser that lets you look at dirs and their subdirs in a single view, a la

![image](https://user-images.githubusercontent.com/2263641/51492315-6ab7e700-1d7f-11e9-9db0-5bd0e5bf5887.png)

That kind of thing would be easy to adapt to this purpose.

**Comment by bollwyvl** [comment]
Well, we have paid the price for forcing schema validation. Well worth it!
Part of the original argument for doing so was to be able to use tools that
can turn a schema/instance into a viewing/editing experience:

https://github.com/mozilla-services/react-jsonschema-form

Perhaps a place to start is an extension that layers that opinion on top,
with an eventual end goal of making it into core.

**Comment by telamonian** [comment]
Aha! Now all of the very carefully formatted JSON makes sense.

[react-jsonschema-form](https://github.com/mozilla-services/react-jsonschema-form) seems neat. I agree that it would be a good fit for the settings editor.

**Comment by bollwyvl** [comment]
Thanks for checking it out! Looking forward to seeing what we get "for
free," as I'm certainly guilty of having written some really nasty schema.

If this does work, it might be useful for other schemaful things... Json
editor, notebook metadata, environment.yml, etc.

what's an example of a setting schema that uses the multiple types
notation? (Sry no grep on phone) It's crazy how many corners of the spec
there are, which highlights how hard it is to build something like this in
the first place.

This is where we get into meta schema stuff where we could validate whether
a schema is representable in the visual editor and disable whole settings
that we can't draw, potentially adding it to developer tooling. I feel like
it wouldn't take too many releases for a user base to pressure an extension
developer to get with the program, while we should be able to audit this
relatively quickly in core if it proves valuable.

At development time there might even be automated suggestions possible to
making something better... In this particular case, it looks like
anyOf/oneOf was just added (unreleased):

https://github.com/mozilla-services/react-jsonschema-form/pull/1133

Which could replace a "type: []" construct without any loss of rigor, and
indeed would provide an opportunity to provide better contextual
documentation. But no doubt other things would show up!

**Comment by bollwyvl** [comment]
@ian-r-rose finally had a lookie-look at your branch: i'd say despite the rough edges, that's pretty good for a first go out! I think once the `anyOf` or `oneOf` stuff lands, it would be reasonable to patch that particular case hot, if it's non-trivial to actually make the upstream config changes. Or, we could endeavor to add support for multi-types. Indeed, some of them don't even make sense:
```json
{
  "type": ["integer", "null"],
  "minimum": 1,
  "maximum": 100
}
```
whereas 
```
{
  "oneOf": [
    {"type": "null"}
    {"type": "integer", "minimum": 1, "maximum": 100}
  ]
}
```
A tree control might _still_ be a better experience, but this is probably a thread worth pulling on!

**Comment by bollwyvl** [comment]
Did some looking at what the next release of rsjf might do for us, and played with style:

![screenshot from 2019-02-03 17-33-00](https://user-images.githubusercontent.com/45380/52183814-c5b6f880-27d9-11e9-81b8-9c911973c924.png)

[The branch](https://github.com/bollwyvl/jupyterlab/tree/settings-form) needs a locally-built version of rsjf `master` next to `jupyterlab`... and i hacked in: 
- a `NullField`
- `peerDependencies` on `react` and `ajv` so we don't get duplicates (and attendant errors)
..But aside from being totally unreproducible for the moment, I'm pretty happy with it.

**Comment by epicfaace** [comment]
@bollwyvl are you able to contribute some of your changes (like a NullField) back upstream to react-jsonschema-form if it makes sense to do so? We'd appreciate it!

**Comment by bollwyvl** [comment]
>  contribute some of your changes

Sure, i'll see what I can do!

> Also, what exactly is the problem with react and ajv dependencies you are seeing?

As of right now, it looks like masters of rsjv is saying
```json
  "peerDependencies": {
    "react": ">=15"
  },
  "dependencies": {
    "ajv": "^6.7.0",
```
and `lab` is saying:
```json
 "react": "~16.4.2",
 "ajv": "^6.5.5",
```
so they _should_ both resolve to the same thing...

At the time (i'd have to dig through my logs of the separate state of two `master` branches :stuck_out_tongue_closed_eyes:), i was getting duplicates of both `react` and `ajv` when doing a `jupyter lab build` (webpack). 

A brief aside:
> > We have a separate command, `jupyter labextension`, that manages versions pretty tightly for the jupyterlab core, but any extension can bring any other `dependencies` it wants along, and not planning carefully can lead to a lot of (not necessarily node-savvy) people doing a lot of building and then serving of an app that, in the case of big dependencies like `ajv` and `react`, multiple versions is not just bloated, but actively dangerous to their stuff working.

Anyhow, so it looks like giving it another go is in order.
